Add `instruction_supported` to `QkTarget`.

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
The following commits add the equivalents to `Target::instruction_supported` to the C API for `Target`.

- `qk_target_instruction_supported`: accepts a gate name (`*mut c_char`), the qargs (`*mut uint32_t`), and the parameters (`*mut QkParam`). It calls upon ``Target::instruction_supported`` as it now supports parameter checks.
- Angle bounds will not be supported in the current version.

### Details and comments
- [x] Add release note.

raynelfss committed:
Initial: Add `instruction_supported` to `QkTarget`.
- Add variants for `reset` and `measure`.
- Add unit tests.

raynelfss committed:
Refactor: Remove `num_qubits` as an argument, use the gate's specifications.

raynelfss committed:
Docs: Add release note

raynelfss committed:
Chore: Remove unused result variable in `create_sample_target`.

raynelfss committed:
Docs: Fix docstring and release note

eliarbel commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -829,6 +829,128 @@ pub unsafe extern "C" fn qk_target_num_instructions(target: *const Target) -> us
+/// # Example
+///
+///     QkTarget *target = qk_target_new(5);
+///     QkTargetEntry *crx_entry = qk_target_entry_new_fixed(QkGate_CRX, *[3.14]);
```
```suggestion
///     QkTargetEntry *crx_entry = qk_target_entry_new_fixed(QkGate_CRX, (double[]){3.14});
```

eliarbel commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -829,6 +829,128 @@ pub unsafe extern "C" fn qk_target_num_instructions(target: *const Target) -> us
+///     qk_target_add_instruction(target, crx_entry);
+///     qk_target_add_instruction(target, QkGate_H, NULL, NULL);
+///
+///     qk_target_instruction_supported(target, QkGate_CRX, [0, 1])
```
```suggestion
///     qk_target_instruction_supported(target, QkGate_CRX, (uint32_t[]){0, 1});
```

eliarbel replied:
Please make sure the example code compiles

eliarbel commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -829,6 +829,128 @@ pub unsafe extern "C" fn qk_target_num_instructions(target: *const Target) -> us
+///
+///     QkTarget *target = qk_target_new(5);
+///     QkTargetEntry *crx_entry = qk_target_entry_new_fixed(QkGate_CRX, *[3.14]);
+///     qk_target_entry_add(crx_entry, NULL, 0, 0.0, 0.1);
```
```suggestion
```
We don't have a `qk_target_entry_add` function....

raynelfss replied:
This is a typo. The correct function should be `qk_target_entry_add_property`.

eliarbel commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -829,6 +829,128 @@ pub unsafe extern "C" fn qk_target_num_instructions(target: *const Target) -> us
+/// Behavior is undefined if ``target`` is not a valid, non-null pointer to a ``QkTarget``.
+///
+/// The ``qargs`` type is expected to be a pointer to an array of ``u32int_t`` where the length
+/// matches the expectation of the gate. If the array is insufficently long the behavior of this
```
```suggestion
/// matches the expectation of the gate. If the array is insufficiently long the behavior of this
```

eliarbel commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -829,6 +829,128 @@ pub unsafe extern "C" fn qk_target_num_instructions(target: *const Target) -> us
+pub unsafe extern "C" fn qk_target_instruction_supported(
+    target: *const Target,
+    operation: StandardGate,
+    qargs: *mut u32,
```
Why are we using `*mut u32` pointers and not `*const u32`?

raynelfss replied:
I'm truly not sure here. I'll make the change promptly.

raynelfss replied:
Addressed in 22a883d

eliarbel commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -829,6 +829,128 @@ pub unsafe extern "C" fn qk_target_num_instructions(target: *const Target) -> us
+///
+/// # Example
+///
+///     QkTarget *target = qk_target_new(5);
```
Does this code compile?

raynelfss replied:
No, it has a typo at the line `qk_target_entry_add` which should be `qk_target_entry_add_property`.

raynelfss replied:
Addressed in 22a883d

eliarbel commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -829,6 +829,128 @@ pub unsafe extern "C" fn qk_target_num_instructions(target: *const Target) -> us
+///
+/// # Example
+///
+///     QkTarget *target = qk_target_new(5);
```
Does this compile?

raynelfss replied:
No, it has some typos as the previous examples. There's also a typo on how `NaN` was written. The correct usage should be `NAN`.

raynelfss replied:
Addressed in 22a883d

eliarbel commented on releasenotes/notes/add-instruction-supported-6323e618f04ae77e.yaml:
```diff
@@ -0,0 +1,41 @@
+  - |
+    Add functions ``qk_target_instruction_supported``, ``qk_target_reset_supported``, and
+    ``qk_target_measure_supported`` to check the compatibility of a ``QkGate``, a :class:`.Reset`
+    instruction, or a :class:`.Measure` with the instance of ``QkTarget``
```
```suggestion
    instruction, or a :class:`.Measure` with the instance of :cpp:class:`QkTarget`.
```
This will probably not make `QkTarget` a real cross reference but at least it'll make look it like one.

raynelfss replied:
I will use `:c:type:` instead, see https://www.sphinx-doc.org/en/master/usage/domains/c.html.

eliarbel commented on releasenotes/notes/add-instruction-supported-6323e618f04ae77e.yaml:
```diff
@@ -0,0 +1,41 @@
+    .. code-block:: c
+      
+      // Create sample target
+      QkTarget *target = qk_target_new(0);
```
I think the current example is overly complex for a release note and the particular new functions. They could be demonstrated with simpler and shorter code. Also, we should make sure the code compiles.

raynelfss replied:
I can remove the overly complex target building step by removing the many arguments I've used. But I'm not sure what else can be done to simplify this. Having at least 2 working gates for the example should be good.

eliarbel replied:
Maybe in this case we should just leave the release note without an example and refer the reader to the function documentation, which includes an example code anyway?

eliarbel reviewed (changes requested):
Thanks @raynelfss for adding this. While reviewing this I had this question: the `instruction_supported` method is very useful for passes and is widely used in our Python and Rust pass implementations. Are there specific use cases in mind for which a C API version of it and its variants would be needed, given that at this stage we are not working towards supporting developing passes in C/C++?

Anway, my main comments are about code examples which currently do not compile. Please see my inline comments.

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -829,6 +829,128 @@ pub unsafe extern "C" fn qk_target_num_instructions(target: *const Target) -> us
+pub unsafe extern "C" fn qk_target_instruction_supported(
+    target: *const Target,
+    operation: StandardGate,
+    qargs: *mut u32,
```
When creating a target we can give `NULL` as `qargs` to indicate a global op. Did you want the same behavior here?

raynelfss replied:
Yes, the behavior here should still supports that based on the docstring and the behavior of the `parse_qargs` helper function.

raynelfss committed:
Chore: Address documentation issues.

Co-authored-by: Eli Arbel <46826214+eliarbel@users.noreply.github.com>

raynelfss committed:
Chore: Fix docstring.

raynelfss commented:
> Are there specific use cases in mind for which a C API version of it and its variants would be needed, given that at this stage we are not working towards supporting developing passes in C/C++?

@eliarbel at the time I began developing this function, I had been asked to think about what functions would someone that's building a transpiler pass from C would use, since this function is essential to checking a backend's compatibility with certain gates and qargs. If we're no longer pursuing this or we're delaying that to other releases, I am more than happy to have this deferred.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into target-c-inst-supported

eliarbel commented:
Quickly going over this again - why not having `qk_instruction_supported` accept an instruction name instead of a `QkGate` param, so it can also cover for the standard instruction cases?

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into target-c-inst-supported

raynelfss committed:
Fix: Merge conflicts.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into target-c-inst-supported

raynelfss committed:
Refactor: Streamline `qk_target_instruction_supported`.
- Accept a C originating string instead of a `QkGate` instance.
- Remove variant methods `qk_target_{measure/reset}_supported`.
- Accept an array of `QkParam` to check for parameters.
- Cleaned up and updated test case.

raynelfss committed:
Fix: Remove python dependency to check numeric value of param.
This commit removes the python dependency when checking if a float originating from the original instruction and a symbol being compared to it have the same value.

raynelfss committed:
Docs: Update release note.

raynelfss committed:
Docs: Add missing free in release note

eliarbel commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -871,6 +872,87 @@ pub unsafe extern "C" fn qk_target_num_instructions(target: *const Target) -> us
+/// @param operation_name The instruction name to check for.
+/// @param qargs The pointer to the array of ``uint32_t`` values to use as
+/// qargs. Can be ``NULL`` if global.
+/// @param params The pointer to the array of ``QkParam`` values as parameters
```
```suggestion
/// @param params A pointer to an array of pointers of ``QkParam`` objects as parameters
```

eliarbel commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -871,6 +872,87 @@ pub unsafe extern "C" fn qk_target_num_instructions(target: *const Target) -> us
+///     // Create a mock target with only a global crx entry
+///     // and 3.14 as its rotation parameter.
+///     QkTarget *target = qk_target_new(5);
+///     QkTargetEntry *crx_entry = qk_target_entry_new_fixed(QkGate_CRX, (QkParam *[]){qk_param_from_double(3.14)});
```
```suggestion
///     QkTargetEntry *crx_entry = qk_target_entry_new_fixed(QkGate_CRX, (double[]){3.14});
```

eliarbel commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -871,6 +872,87 @@ pub unsafe extern "C" fn qk_target_num_instructions(target: *const Target) -> us
+///
+///     // Check if target is compatible with a "crx" gate
+///     // at [0, 1] with 3.14 rotation.
+///     QkParam params[1] = {qk_param_from_double(3.14),};
```
```suggestion
///     QkParam *params[1] = {qk_param_from_double(3.14)};
```

eliarbel commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -871,6 +872,87 @@ pub unsafe extern "C" fn qk_target_num_instructions(target: *const Target) -> us
+///     // Check if target is compatible with a "crx" gate
+///     // at [0, 1] with 3.14 rotation.
+///     QkParam params[1] = {qk_param_from_double(3.14),};
+///     qk_target_instruction_supported(target, "crx", (uint32_t *){0, 1}, params);
```
```suggestion
///     qk_target_instruction_supported(target, "crx", (uint32_t []){0, 1}, params);
```

eliarbel commented on crates/cext/src/transpiler/target.rs:
````diff
@@ -871,6 +872,87 @@ pub unsafe extern "C" fn qk_target_num_instructions(target: *const Target) -> us
+///     // at [0, 1] with 3.14 rotation.
+///     QkParam params[1] = {qk_param_from_double(3.14),};
+///     qk_target_instruction_supported(target, "crx", (uint32_t *){0, 1}, params);
+/// ```
````
I think we should also include these at the end:
```c
qk_param_free(params[0]);
qk_target_free(target);
```

raynelfss replied:
Added in 0aee96540cb47619ff9822258f6672b9dba3d992

eliarbel commented on test/c/test_target.c:
```diff
@@ -544,6 +545,98 @@ static int test_target_update_instruction(void) {
+                printf("This target did not correctly demonstrate compatibility with %s and qargs "
+                       "[%d]",
+                       gate_names[j], i);
+                result = RuntimeError;
```
Should we use `EqualityError` instead throughout the checks in the function? It seems that we use `EqualityError` for unexpected test outcomes and `RuntimeError` for failures which are not part of the test logic, but I may be over-generalizing.

raynelfss replied:
You are right, the reason I was choosing runtime over equality was the error being in the implementation rather than the result itself. But it doesn't make things unrecoverable in this case but rather it just provides the wrong result.

eliarbel replied:
I think the other return values should be changed as well.

raynelfss replied:
Oh, I must have missed the others. Done in a11d54f0fd53d219a53f6555309c5cede843337f

eliarbel commented on test/c/test_target.c:
```diff
@@ -544,6 +545,98 @@ static int test_target_update_instruction(void) {
+                               qk_param_from_double(3.14),
+                           },
+                           NULL};
+    for (uint32_t i = 0; i < 5; i++) {
```
Small nit 1:
```suggestion
    for (uint32_t qubit = 0; qubit < 5; qubit++) {
```

eliarbel commented on test/c/test_target.c:
```diff
@@ -544,6 +545,98 @@ static int test_target_update_instruction(void) {
+        uint32_t qargs[1] = {i};
+        bool should_be_true = i < 4;
+
+        for (int j = 0; j < 5; j++) {
```
Small nit 2 😸 

```suggestion
        for (int gate = 0; gate < 5; gate++) {
```

eliarbel commented on test/c/test_target.c:
```diff
@@ -544,6 +545,98 @@ static int test_target_update_instruction(void) {
+                   i);
+            result = RuntimeError;
+            goto cleanup;
+        }
```
Can `reset` just be part of `gate_names`?

raynelfss replied:
Added in 0aee96540cb47619ff9822258f6672b9dba3d992

eliarbel commented on test/c/test_target.c:
```diff
@@ -544,6 +545,98 @@ static int test_target_update_instruction(void) {
+                           (QkParam *[]){
+                               qk_param_from_double(3.14),
+                           },
+                           NULL};
```
A minor style suggestion (alternatively, I miss a ternary operator in Rust and here is an opportunity to use one 😉 ): 
```c
QkParam *param[1] = {qk_param_from_double(3.14)};
```
Followed by: 
```c
 if (!(qk_target_instruction_supported(sample_target, gate_names[j], qargs, j != 3 ? NULL : params)
```

raynelfss replied:
Added in 0aee96540cb47619ff9822258f6672b9dba3d992

eliarbel commented on releasenotes/notes/add-instruction-supported-6323e618f04ae77e.yaml:
```diff
@@ -0,0 +1,44 @@
+---
+features_c:
+  - |
+    Add function ``qk_target_instruction_supported`` to check the compatibility of any gate
```
```suggestion
    Add function :c:func:`qk_target_instruction_supported` to check the compatibility of any gate
```

eliarbel commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -871,6 +872,87 @@ pub unsafe extern "C" fn qk_target_num_instructions(target: *const Target) -> us
+        CStr::from_ptr(operation_name.cast())
+            .to_str()
+            .expect("Error extracting gate name from target.")
+    };
```
Why is `cast()` needed here? 
Maybe also narrow `unsafe` scope: 
```suggestion
    let name = unsafe { CStr::from_ptr(operation_name) }
        .to_str()
        .expect("Error extracting gate name from target.");
```

eliarbel commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -871,6 +872,87 @@ pub unsafe extern "C" fn qk_target_num_instructions(target: *const Target) -> us
+
+    // SAFETY: Per documentation, the params pointer points to an appropiately allocated
+    // array of `QkParam` objects.
+    let params: Vec<_> = if params.is_null() {
```
`Vec<_>` could be dropped

raynelfss replied:
Dropped in 0aee96540cb47619ff9822258f6672b9dba3d992

eliarbel commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -871,6 +872,87 @@ pub unsafe extern "C" fn qk_target_num_instructions(target: *const Target) -> us
+    let name = unsafe {
+        CStr::from_ptr(operation_name.cast())
+            .to_str()
+            .expect("Error extracting gate name from target.")
```
```suggestion
            .expect("Error extracting gate name from operation_name.")
```

raynelfss committed:
Apply suggestions from code review

Co-authored-by: Eli Arbel <46826214+eliarbel@users.noreply.github.com>

raynelfss committed:
Apply suggestions from code review

Co-authored-by: Eli Arbel <46826214+eliarbel@users.noreply.github.com>

raynelfss committed:
Apply suggestions from code review

Co-authored-by: Eli Arbel <46826214+eliarbel@users.noreply.github.com>

raynelfss committed:
Merge branch 'main' into target-c-inst-supported

raynelfss committed:
Fix: Address review comments

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into target-c-inst-supported

eliarbel commented on test/c/test_target.c:
```diff
@@ -552,6 +553,90 @@ static int test_target_update_instruction(void) {
+            if (!(qk_target_instruction_supported(sample_target, gate_names[gate], qargs,
+                                                  gate != 3 ? NULL : rz_params) ==
+                  should_be_true) &&
+                gate != 1) {
```
This doesn't properly validate that `y` is supported as a global instruction since it's short circuiting on `gate == 1`. 
We can refactor the condition to something like this:

```suggestion
            if (qk_target_instruction_supported(sample_target, gate_names[gate], qargs,
                                                  gate != 3 ? NULL : rz_params) !=
                  (should_be_true || gate == 1)) {
```

eliarbel commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -942,6 +942,89 @@ pub unsafe extern "C" fn qk_target_num_instructions(target: *const Target) -> us
+        return false;
+    };
+
+    // SAFETY: Per documentation, the params pointer points to an appropiately allocated
```
```suggestion
    // SAFETY: Per documentation, the params pointer points to an appropriately allocated
```

eliarbel commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -942,6 +942,89 @@ pub unsafe extern "C" fn qk_target_num_instructions(target: *const Target) -> us
+    };
+
+    // SAFETY: Per documentation, the params pointer points to an appropiately allocated
+    // array of `QkParam` objects.
```
We should make the safety comments about the `params` parameter clearer about not only the array itself being properly allocated but also about its entries being valid pointers to `QkParam` objects.

raynelfss replied:
Would something like this suffice?

```rs
    // SAFETY: Per documentation, the params argument points to an appropriately allocated
    // array of valid pointers to `QkParam` objects.
```

eliarbel replied:
LGTM

raynelfss committed:
Apply suggestions from code review

Co-authored-by: Eli Arbel <46826214+eliarbel@users.noreply.github.com>

raynelfss committed:
Fix: Address more review comments

raynelfss committed:
Merge branch 'main' into target-c-inst-supported

raynelfss committed:
Fix: Change `RuntimeError` to
`EqualityError`

eliarbel reviewed (approved):
Thanks Ray for all the work here, it looks great now!
